### DISCUSSION TITLE
Version App MacOS how to ?

### DISCUSSION THREAD

**Discussion by TNNSLAB** [pyinstaller/pyinstaller#8033] [Help]
I want to add the version to the Mac Os application...failing...

What am I doing wrong ? I need to build the app into one file and add the version to it...

here's my app folder and it's contains : 
```
version.txt
transfers.py
vendor.py
movenet_lightning_fp16.tflite
requirements.txt
model_crop_usa.py
model_crop.py
icon2.png
grey_background.py
download.py
dist
crop.py
build
app.spec
app.py
```

my **app.spec** looks like this : 

```
# -*- mode: python ; coding: utf-8 -*-```
import sys
sys.setrecursionlimit(5000)

a = Analysis(
    ['app.py'],
    pathex=['/Users/adrian/Desktop/Imaging_App'],
    binaries=[],
    datas = [('/Users/adrian/Desktop/Imaging_App/movenet_lightning_fp16.tflite', '.'),
         ('/Users/adrian/Desktop/Imaging_App/icon2.png', '.')],
    hiddenimports=[],
    hookspath=[],
    hooksconfig={},
    runtime_hooks=[],
    excludes=[],
    noarchive=False
)
pyz = PYZ(a.pure)

exe = EXE(
    pyz,
    a.scripts,
    [],
    exclude_binaries=True,
    name='app',
    debug=False,
    bootloader_ignore_signals=False,
    strip=False,
    upx=True,
    console=True,
    disable_windowed_traceback=False,
    argv_emulation=False,
    target_arch=None,
    codesign_identity=None,
    entitlements_file=None,
)
coll = COLLECT(
    exe,
    a.binaries,
    a.datas,
    strip=False,
    upx=True,
    upx_exclude=[],
    name='app',
)
```

Here's my **version.txt** file : 
```
# UTF-8
VSVersionInfo(
        ffi=FixedFileInfo(
        filevers=(ProductVersions),
        prodvers=(ProductVersions),
        mask=0x3f,
        flags=0x0,
        OS=0x40004,
        fileType=0x1,
        subtype=0x0,
        date=(0, 0)),
        kids=[StringFileInfo([StringTable(
        u'040904B0',
        [StringStruct(u'FileDescription', u'xyz'),
        StringStruct(u'FileVersion', u'2.3.0.0'),
        StringStruct(u'InternalName', u'xyz'),
        StringStruct(u'LegalCopyright', u'Copyright Adrian'),
        StringStruct(u'OriginalFilename', u'xyz'),
        StringStruct(u'ProductName', u'xyz'),
        StringStruct(u'ProductVersion', u'2.4.0.0'),
        StringStruct(u'Language', u'Language Neutral'),
        StringStruct(u'LegalTrademarks', u'xyz')])]), 
        VarFileInfo([VarStruct(u'Translation', [1033, 1200])])]
    )
```

**Comment by rokm** [comment]
> What am I doing wrong ? 

That **version.txt** is Windows-specific, and will work only on Windows (and even there, only if you pass `version='version.txt'` as an argument to `EXE()` in the spec file).

> I need to build the app into one file and add the version to it...

On macOS, POSIX builds don't have any version attribute, only .app bundles do. 

So you'd need to generate a .spec template with added `--windowed` flag, or manually add the following `BUNDLE()` statement to the existing spec:

``` 
...
coll = COLLECT(
    ...
)

app = BUNDLE(
    coll,
    name='app.app',
    icon=None,
    bundle_identifier=None,  # <- you should set your bundle identifier here
    version='2.3.0',
)
```

The specified version should be written to the .app bundle's `Info.plist` file.

**Reply by TNNSLAB** [reply]
@rokm stupid question...please don't judge me....how i can set the bundle identifier ?

**Reply by rokm** [reply]
> @rokm stupid question...please don't judge me....how i can set the bundle identifier ?

`bundle_identifier='my.bundle.identifier'`

See [here](https://github.com/sabnzbd/sabnzbd/blob/08e61ecf19499aa1b31740f5be160186191479d5/builder/SABnzbd.spec#L152-L180) for a more detailed example (note that they set version via `info_plist`, though).

**Reply by TNNSLAB** [reply]
that looks like something that might work ! 
one last stupid question from me... what is the correct command to build onefile ? 

```pyinstaller -onefile app.spec ``` would something like this work ?

**Reply by rokm** [reply]
Onefile vs onedir is determined by the spec file and its contents. So if you want to switch between the two and don't know the exact details, you should generate new spec file from the template (`pyinstaller [any-other-arguments] --onefile app.py`  or `pyinstaller [any-other-arguments] --onedir app.py`). 

That said, if you want to generate an .app bundle, you should not be using onefile mode anyway (especially if you plan to do any code-signing).

**Reply by TNNSLAB** [reply]
Ohhh ok .... 

pysintaller app.spec 

and that generates two types of applications: 

<img width="636" alt="Screenshot 2023-10-24 at 11 40 56" src="https://github.com/pyinstaller/pyinstaller/assets/46636486/12a012ee-b507-433d-827f-b95dd96d1197">

The exe filetype application launches and works properly...where app.app tries to launch and it closes automatically...

Sorry again for the dummy questions...

**Reply by rokm** [reply]
If it closes automatically, it most likely crashed due to unhandled exception. So your code is doing something that violates the restricted environment in which .app bundles run - like for example trying to write to application's directory (or the current working directory).

If you want to figure out what went wrong, you'll have to (temporarily) wrap whole code in `app.py` into a try/catch block, and write the exception information to a file in user-writable location (like the home directory). See here for an example: https://github.com/pyinstaller/pyinstaller/issues/6992#issuecomment-1198434935

**Reply by TNNSLAB** [reply]
ok I will try that...could that be also because of Code Signing ? 

I'm not doing anything there...

**Reply by rokm** [reply]
If you're not signing the bundle with actual code-signing identity, then the problem is likely unrelated to code signing.

**Reply by TNNSLAB** [reply]
it weird, 'cause the app.exec inside the app.app works when i open it...but the app.app it starts to open than closes....

<img width="627" alt="Screenshot 2023-10-24 at 11 57 10" src="https://github.com/pyinstaller/pyinstaller/assets/46636486/fdc4983f-c756-4063-b36a-086c9c9a9413">

**Reply by rokm** [reply]
Then your code likely depends on some functionality that is available to regular POSIX executables but not to .app bundles (e.g., .app bundles run with reduced set of environment variables, because shell is not initialized). Without seeing your code, it will probably be fastest if you try to catch and log the exception.
